solr/core/src/java/org/apache/solr/search/facet/FieldUtil.java

```diff
@@ -44,27 +48,49 @@ public static SortedDocValues getSortedDocValues(SolrIndexSearcher searcher, Str
 
   public static SortedDocValues getSortedDocValues(
       QueryContext context, SchemaField field, QParser qparser) throws IOException {
-    SortedDocValues si =
-        context.searcher().getSlowAtomicReader().getSortedDocValues(field.getName());
-    // if (!field.hasDocValues() && (field.getType() instanceof StrField || field.getType()
-    // instanceof TextField)) {
-    // }
-
-    return si == null ? DocValues.emptySorted() : si;
+    var reader = context.searcher().getSlowAtomicReader();
+    var dv = reader.getSortedDocValues(field.getName());
+    checkDvType(dv, field, reader);
+    return dv == null ? DocValues.emptySorted() : dv;
   }
 
   public static SortedSetDocValues getSortedSetDocValues(
       QueryContext context, SchemaField field, QParser qparser) throws IOException {
-    SortedSetDocValues si =
-        context.searcher().getSlowAtomicReader().getSortedSetDocValues(field.getName());
-    return si == null ? DocValues.emptySortedSet() : si;
+    var reader = context.searcher().getSlowAtomicReader();
+    var dv = reader.getSortedSetDocValues(field.getName());
+    checkDvType(dv, field, reader);
+    return dv == null ? DocValues.emptySortedSet() : dv;
   }
 
   public static NumericDocValues getNumericDocValues(
       QueryContext context, SchemaField field, QParser qparser) throws IOException {
-    SolrIndexSearcher searcher = context.searcher();
-    NumericDocValues si = searcher.getSlowAtomicReader().getNumericDocValues(field.getName());
-    return si == null ? DocValues.emptyNumeric() : si;
+    var reader = context.searcher().getSlowAtomicReader();
+    var dv = reader.getNumericDocValues(field.getName());
+    checkDvType(dv, field, reader);
+    return dv == null ? DocValues.emptyNumeric() : dv;
+  }
+
+  private static void checkDvType(Object dv, SchemaField field, LeafReader reader) {
```

Comment:
It's not clear why to have this 20 lines. If we remove it, I think it just fails at class cast somewhere later. Why ISE is better than ClassCastException later? It's  just not obvious.

Reply:
There's wasn't a class cast; there was empty/no data which ultimately leads to no facet results.  That's *worse* than throwing an exception (of whatever type; no strong opinion from me).  This is my primary addition to the PR; don't *just* fix the wrong algorithm choice, but ensure we throw an exception in cases that would previously silently produce nothing.